src/app/qgsmaptoolselectutils.cpp

```diff
@@ -335,3 +337,193 @@ QgsFeatureIds QgsMapToolSelectUtils::getMatchingFeatures( QgsMapCanvas *canvas,
   return newSelectedFeatures;
 }
 
+
+QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::QgsMapToolSelectMenuActions(
+  QgsMapCanvas *canvas,
+  QgsVectorLayer *vectorLayer,
+  QgsVectorLayer::SelectBehavior behavior,
+  QObject *parent ):
+  QObject( parent ),
+  mCanvas( canvas ),
+  mVectorLayer( vectorLayer ),
+  mBehavior( behavior )
+{
+  connect( mVectorLayer, &QgsMapLayer::destroyed, this, &QgsMapToolSelectMenuActions::onLayerDestroyed );
+}
+
+QList<QAction *> QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::actions( const QgsFeatureIds &featureCanditateIds )
+{
+  QList<QAction *> actionsList;
+
+  QString beginningOfText;
+
+  QgsFeatureIds effectiveFeatureIds = featureCanditateIds;
+
+  switch ( mBehavior )
+  {
+    case QgsVectorLayer::SetSelection:
+      beginningOfText = tr( "Select " );
+      break;
+    case QgsVectorLayer::AddToSelection:
+    {
+      beginningOfText = tr( "Add " );
+      QgsFeatureIds existingSelection = mVectorLayer->selectedFeatureIds();
+      for ( const QgsFeatureId &selected : featureCanditateIds )
+      {
+        if ( existingSelection.contains( selected ) )
+          effectiveFeatureIds.remove( selected );
+      }
+    }
+    break;
+
+    case QgsVectorLayer::IntersectSelection:
+    {
+      beginningOfText = tr( "Intersect " );
+      QgsFeatureIds existingSelection = mVectorLayer->selectedFeatureIds();
+      for ( const QgsFeatureId &selected : featureCanditateIds )
+      {
+        if ( !existingSelection.contains( selected ) )
+          effectiveFeatureIds.remove( selected );
+      }
+    }
+    break;
+    case QgsVectorLayer::RemoveFromSelection:
+    {
+      beginningOfText = tr( "Remove " );
+      QgsFeatureIds existingSelection = mVectorLayer->selectedFeatureIds();
+      for ( const QgsFeatureId &selected : featureCanditateIds )
+      {
+        if ( !existingSelection.contains( selected ) )
+          effectiveFeatureIds.remove( selected );
+      }
+    }
+    break;
+  }
+
+  if ( !effectiveFeatureIds.isEmpty() )
+  {
+    if ( effectiveFeatureIds.size() > 1 )
+    {
+      QAction *actionAll = new QAction( beginningOfText + tr( "All Features (%1)" ).arg( effectiveFeatureIds.count() ), this );
+      connect( actionAll, &QAction::triggered, this, &QgsMapToolSelectMenuActions::selectFeature );
+      connect( actionAll, &QAction::hovered, this, &QgsMapToolSelectMenuActions::highLightFeatures );
+      QVariantList list;
+      for ( const QgsFeatureId &id : effectiveFeatureIds )
```

Comment:
```suggestion
      for ( QgsFeatureId id : qgis::as_const( effectiveFeatureIds ) )
```

Reply:
Can you make sure all feature ids are iterated

1. by value (it's a simple 64 bit int, referencing is not worth it)
2. the container is const